### DISCUSSION TITLE
stories_things/basic_content/macro/

### DISCUSSION THREAD

**Discussion by giscus** [Light-City/light-comments#16] [Announcements]
# stories_things/basic_content/macro/

Stories About C Plus Plus

https://light-city.github.io/stories_things/basic_content/macro/

**Comment by BobLiu21** [comment]
最后这个其实可以直接用一个大括号就好了。

**Reply by Light-City** [reply]
是的，没毛病

**Comment by Dusk-Hermit** [comment]
乐了，看了你的笔记和你抄的那篇学习笔记，发现你偷理解没偷明白。他说用宏用大括号包起来是不对的，多了一个分号，你把这一点改成了`（虽然编译运行没问题）`。实际上看原贴https://www.pixelstech.net/article/1390482950-do-%7B-%7D-while-%280%29-in-macros 末尾提到，if else的时候，宏替换残留的分号会导致语法问题

**Reply by Dusk-Hermit** [reply]
最搞的是，你参考的原贴也是原封不动翻译它的原贴，只补充一点点代码例子的……

**Reply by BobLiu21** [reply]
朋友，首先你提出问题我觉得挺好的，但是我觉得你可能误会了，最后一个是2.4，楼主的意思是缩小作用域，我说不用do。。while，直接用大括号，这个跟宏有什么关系呢？你觉得呢？

**Reply by Dusk-Hermit** [reply]
你能收到讯息应该是因为你在这回复从而自动订阅了。但我指的是2.1的文章细节，不是你提出的问题。

**Reply by Light-City** [reply]
友好的交流能够帮助大家互相学习，上来就是“乐了，看了你的笔记和你抄的那篇学习笔记，发现你偷理解没偷明白。”

这是在嘲讽我？还是讥讽我？

说到这个问题上来，你发的博客链接我看都没看过怎么炒，而且我这里表达的意思你明白了吗？

我还是给你解释一下：

```
f(a>0)
    fun();
// 宏展开
if(a>0)
{
    f1();
    f2();
};
```

这里是想表达：当一个func()的内容用括号括起来的时候，由于按照用户编写的习惯，可能后面会加一个分号

```
#define fun() {f1();f2();}
f(a>0)
    fun();
```

那么会展开为

```
if(a>0)
{
    f1();
    f2();
};
```
实际上这个分号是多余的，并不需要，所以上面说了写的语法不太对，改为func()就可以了，后面还加了注释去说明(虽然编译运行没问题，正常没分号)。

**Reply by Light-City** [reply]
f(a > 0) 是 if (a > 0)少写一个if
